espnet/utils/spec_augment.py

```diff
@@ -0,0 +1,456 @@
+# -*- coding: utf-8 -*-
+
+"""
+This implementation is modified from https://github.com/zcaceres/spec_augment
+
+MIT License
+
+Copyright (c) 2019 Zach Caceres
+
+Permission is hereby granted, free of charge, to any person obtaining a copy
+of this software and associated documentation files (the "Software"), to deal
+in the Software without restriction, including without limitation the rights
+to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+copies of the Software, and to permit persons to whom the Software is
+furnished to do so, subject to the following conditions:
+
+The above copyright notice and this permission notice shall be included in all
+copies or substantial portions of the Software.
+
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETjjHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+SOFTWARE.
+"""
+
+import random
+
+import torch
+
+
+def specaug(spec, W=5, F=30, T=40, num_freq_masks=2, num_time_masks=2, replace_with_zero=False):
+    """SpecAugment
+
+    Reference: SpecAugment: A Simple Data Augmentation Method for Automatic Speech Recognition
+        (https://arxiv.org/pdf/1904.08779.pdf)
+
+    This implementation modified from https://github.com/zcaceres/spec_augment
+
+    :param torch.Tensor spec: input tensor with the shape (T, dim)
+    :param int W: time warp parameter
+    :param int F: maximum width of each freq mask
+    :param int T: maximum width of each time mask
+    :param int num_freq_masks: number of frequency masks
+    :param int num_time_masks: number of time masks
+    :param bool replace_with_zero: if True, masked parts will be filled with 0, if False, filled with mean
+    """
+    return time_mask(
+        freq_mask(time_warp(spec, W=W),
+                  F=F, num_masks=num_freq_masks, replace_with_zero=replace_with_zero),
+        T=T, num_masks=num_time_masks, replace_with_zero=replace_with_zero)
+
+
+def time_warp(spec, W=5):
+    """Time warping
+
+    :param torch.Tensor spec: input tensor with shape (T, dim)
+    :param int W: time warp parameter
+    """
+    spec = spec.unsqueeze(0)
+    spec_len = spec.shape[1]
+    num_rows = spec.shape[2]
+    device = spec.device
+
+    y = num_rows // 2
+    horizontal_line_at_ctr = spec[0, :, y]
+    assert len(horizontal_line_at_ctr) == spec_len
+
+    point_to_warp = horizontal_line_at_ctr[random.randrange(W, spec_len - W)]
```

Comment:
I am not so sure about this part. It looks very strange because point_to_warp is a value randomly selected from center line but it looks like a coordinate is needed (src and dest pts). Someone actually asked in [original repo](https://github.com/zcaceres/spec_augment/issues/3).

Without understand of "warping" operation I cannot make a conclusion now.

Reply:
According to specaug paper time warping is not that important so maybe we can skip this part right now. Any comments are welcome.

Reply:
I have not read the details of the paper and the notebook yet. If the time warping is similar to speed perturbation, some tasks (CSJ, tedlium, etc) might prefer it, and the other (librispeech?) might not.

Reply:
After I read the paper and `sparse_image_warp`, I suspect that it could be much simpler using Pillow (torchvision). It is getting faster (32ms -> 2ms) and more stable (I found the ref impl got warning sometimes).

See `time_warp_pil` in https://gist.github.com/ShigekiKarita/d6f6a77c4c9b929b3a5fdb10ed1a70a9

Reply:
I feel that the time warping is similar to [RandomResizedCrop](https://pytorch.org/docs/stable/torchvision/transforms.html#torchvision.transforms.RandomResizedCrop) rather than speed perturbation. It is popular data augumentation in ImageNet.

Reply:
That tilted spectrogram in the ref impl is already reported but I do not think this is the problem of window size. https://github.com/zcaceres/spec_augment/issues/5